Update spell check GHA and use_spell_check() with configurable inputs

Bai-Li-NOAA opened the pull request:
This PR addresses issue #167:
* feat(spell-check.yml): allow users to run spell check on specific file types
  * supported file types: .cpp, .hpp, .md, .R, .Rmd, .txt, and .yml
  * users can specify whether spelling errors should be reported as a warning or an error
* feat(use_spell_check()): add extra parameters to control spell check report level
* test(test-use_r_workflows.R): add additoinal tests and snapshot files for use_spell_check()
* fix(typos): use spelling::spell_check_files() to detect and correct typos

The GHA has been tested with different settings: 1) it [fails](https://github.com/nmfs-ost/ghactions4r/actions/runs/15715925275/job/44285517671) when `full_check_report_level` is set to `error` and 2) it [passes](https://github.com/nmfs-ost/ghactions4r/actions/runs/15716027458/job/44285858793) when `full_check_report_level` is set to `warning`.

@k-doering-NOAA, sorry if your inbox was flooded with failed GHA notifications today. 😅 I finally got the workflow running. Feel free to suggest better names if any of the current ones seem unclear. Also, let me know if you'd like to add any of the [new words](https://github.com/nmfs-ost/ghactions4r/actions/runs/15715925275/job/44285517671#step:6:25) to the WORDLIST.

Bai-Li-NOAA committed:
Update spell check GHA and use_spell_check() with configurable inputs

* feat(spell-check.yml): allow users to run spell check on specific file
  types
  * supported file types: .cpp, .hpp, .md, .qmd, .R, .Rmd, .txt, and .yml
  * users can specify whether spelling errors should be reported as a
    warning or an error
* feat(use_spell_check()): add extra parameters to control spell check
  report level
* test(test-use_r_workflows.R): add additoinal tests and snapshot files
  for use_spell_check()
* fix(typos): use spelling::spell_check_files() to detect and correct
  typos

k-doering-NOAA commented:
@Bai-Li-NOAA thanks so much for working on this! Sorry for the delay. I scheduled some time on 6/23 to work on this.

k-doering-NOAA commented on R/use_r_workflows.R:
```diff
@@ -428,15 +428,59 @@ use_build_pkgdown <- function(workflow_name = "call-build-pkgdown.yml", addition
+
+  # Remind users that full_check_report_level is set to 'warning' by default if users 
+  # don't specify full_check_report_level
+  if (use_full_spell_check == TRUE & length(full_check_report_level) > 1) {
```
@Bai-Li-NOAA , I think either & or && work in this case, so up to you if you want to leave or change.

Bai-Li-NOAA replied:
Great. I will leave it as is then.

k-doering-NOAA commented on .github/workflows/spell-check.yml:
```diff
@@ -1,6 +1,17 @@
+      type: boolean
+      default: false
+    full_check_report_level:
+      description: 'The message level to report if spelling issues are found during the full spell check'
```
Perhaps the description should include the possible values (I'm assuming 'warning' and 'error'?)

Bai-Li-NOAA replied:
Great suggestion. Done!

k-doering-NOAA commented on inst/templates/call-calc-cov-summaries.yml:
```diff
@@ -16,7 +16,7 @@ on:
       - main
       
-# limits permisions to write, as the common use case is for pull requests;
+# limits permissions to write, as the common use case is for pull requests;
```
Haha, thanks spell check files!! :)

k-doering-NOAA commented on .github/workflows/spell-check.yml:
```diff
@@ -26,13 +37,53 @@ jobs:
       - name: Install spelling
         run: Rscript -e 'install.packages("spelling")'
 
-      - name: spell check
```
I think a conditional statement is missing under name, perhaps:
```yml
 if: inputs.use_full_spell_check == false
``` 

(based on https://github.com/nmfs-ost/ghactions4r/blob/a9bf1dd9467cfb9b210f88de7f9b07750e761f8c/.github/workflows/create-cov-badge.yml#L47)

Bai-Li-NOAA replied:
Sorry for the delay getting back to this PR. I think the GHA would always run `results <- spelling::spell_check_package()` because the additional checks don’t cover all the file types that `spell_check_package()` does, for example, Sweave vignettes. So it’s not a replacement for `spell_check_package()`.

To make this clearer, I’ve updated the input name from `use_full_spell_check` to `spell_check_additional_files`. I also added code in the GHA log to show how to update the WORDLIST after finding more words in the additional check.

k-doering-NOAA replied:
Oh, of course it should, thank you for explaining!

k-doering-NOAA reviewed (changes requested):
Hi Bai, thanks so much for working on this!

I made a few suggestions, but the most important one is I think there is a conditional statement missing in the yml - it is needed so that spell_check_package() will be skipped when using use_full_spell_check == true.

The R function `use_spell_check()` with the new options worked great for me, and the new tests and snapshots also look good!

Once a conditional statement is added, I'm happy to try out the workflows using the full_spell_check option on SSMSE, I made a branch to try them out: https://github.com/nmfs-ost/SSMSE/tree/test-spell-check

P.S. I found the AI summary useful - although of course your summary was just as good :)

Bai-Li-NOAA commented:
@k-doering-NOAA, thanks for the review. I will work on addressing the suggestions and let you know once they are incorporated.

I also have two quick questions:
Would you like to use full spell check in this repo? And if so, should the GHA report spell check issues as errors or warnings? If you'd prefer to treat them as errors, I can go ahead and add the [new words](https://github.com/nmfs-ost/ghactions4r/actions/runs/15715925275/job/44285517671#step:6:25) to the WORDLIST.

k-doering-NOAA commented:
@Bai-Li-NOAA, yes, let's use the full check in this repo, and make it error (I will never use the info if it is warning only 🤣 ) - I took a brief glance at the new words and some of them I thought might be typos - but I didn't take a deep look. If you did already, though, feel free to add the wordlist!

Bai-Li-NOAA commented:
@k-doering-NOAA, I made some changes following the suggestions in the PR. New changes can be found [here](https://github.com/nmfs-ost/ghactions4r/pull/209/commits/740d2d6c6b692161f208bbcd6bec3843f3ef5cb7). Let me know if you have additional questions or comments!

k-doering-NOAA commented on .github/workflows/call-spell-check.yml:
```diff
@@ -18,4 +18,7 @@ permissions: {}
+    uses: nmfs-ost/ghactions4r/.github/workflows/spell-check.yml@main
+    with:
+      spell_check_additional_files: true
+      spell_check_report_level: "error"
```
Just noting it should also work without quotations around "error"

Bai-Li-NOAA replied:
Thanks. I removed the quotations.

k-doering-NOAA reviewed (approved):
Thanks @Bai-Li-NOAA - I tested  with SSMSE and the additional spell check option worked well. I like the new name for `spell_check_additional_files`!

A few lingering questions/suggestions are below. But, I think it is up to you on how to deal with them, if at all.

- Should  `spell_check_report_level` apply to the spell check using spelling::spell_check_package() step, too? And if it does apply to both, should error actually be the default? I kind of am thinking "yes" to both of these, but I will leave it up to you :) Thanks for developing this!
- What is the workflow for adding the additional spell check file words to the wordlist? is there an automated way to do it, like with spellling::update_wordlist()? Maybe we can document this for users

Bai-Li-NOAA commented:
@k-doering-NOAA, for the lingering questions/suggestions: 
- When `spelling::spell_check_package()` detects spelling errors, the GHA will report it as an error. Based on my observations in [FIMS](https://github.com/NOAA-FIMS/FIMS/actions/runs/16170004844/job/45640888175#step:5:18), the `spell_check_report_level` setting only applies to checks on additional files. I chose `warning` as the default value because I’m not sure if other packages are ready to treat spelling errors in additional files as `errors`, especially without ignoring all the extra words. For example, FIMS is currently doing a soft check internally and doesn’t plan to ignore all words from the additional check, as the list could get very long. **In short, I’d like to test this on FIMS and a few other packages first before changing the default behavior to error on additional files.** 
- I haven’t found a single function that can automatically update the WORDLIST after checking additional files. But, I’ve included four lines of code that can help add these additional words to the WORDLIST, and they’ll be printed in the GHA logs. I also plan to open an issue in the {spelling} package repo to see if this could be a feature worth adding.

Bai-Li-NOAA commented:
@k-doering-NOAA, turns out my memory was off. I checked with Kelli, and we can update the WORDLIST in FIMS even if it's long. I’ve updated the default `spell_check_report_level` to `error` and merged the PR.
